zcash_client_sqlite/src/wallet/init/migrations/orchard_shardtree.rs

```diff
@@ -0,0 +1,214 @@
+                LEFT OUTER JOIN orchard_tree_shards prev_shard
+                    ON shard.shard_index = prev_shard.shard_index + 1
+                -- Join with scan ranges that overlap with the subtree's involved blocks.
+                INNER JOIN scan_queue ON (
+                    subtree_start_height < scan_queue.block_range_end AND
+                    (
+                        scan_queue.block_range_start <= shard.subtree_end_height OR
+                        shard.subtree_end_height IS NULL
+                    )
+                )",
+            16, // ORCHARD_SHARD_HEIGHT is only available when `feature = "orchard"` is enabled.
+            16, // ORCHARD_SHARD_HEIGHT is only available when `feature = "orchard"` is enabled.
+            u32::from(self.params.activation_height(NetworkUpgrade::Nu5).unwrap()),
+        ))?;
+
+        transaction.execute_batch(&format!(
+            "CREATE VIEW v_orchard_shard_unscanned_ranges AS
+                WITH wallet_birthday AS (SELECT MIN(birthday_height) AS height FROM accounts)
+                SELECT
+                    shard_index,
+                    start_position,
+                    end_position_exclusive,
+                    subtree_start_height,
+                    subtree_end_height,
+                    contains_marked,
+                    block_range_start,
+                    block_range_end,
+                    priority
+                FROM v_orchard_shard_scan_ranges
+                INNER JOIN wallet_birthday
+                WHERE priority > {}
+                AND block_range_end > wallet_birthday.height;",
+            priority_code(&ScanPriority::Scanned),
+        ))?;
+
+        transaction.execute_batch(
+            "CREATE VIEW v_orchard_shards_scan_state AS
+            SELECT
+                shard_index,
+                start_position,
+                end_position_exclusive,
+                subtree_start_height,
+                subtree_end_height,
+                contains_marked,
+                MAX(priority) AS max_priority
+            FROM v_orchard_shard_scan_ranges
+            GROUP BY
+                shard_index,
+                start_position,
+                end_position_exclusive,
+                subtree_start_height,
+                subtree_end_height,
+                contains_marked;",
+        )?;
+
+        // Treat the current best-known chain tip height as the height to use for Orchard
+        // initialization, bounded below by NU5 activation.
+        if let Some(orchard_init_height) = scan_queue_extrema(transaction)?.and_then(|r| {
+            self.params
+                .activation_height(NetworkUpgrade::Nu5)
+                .map(|orchard_activation| std::cmp::max(orchard_activation, *r.end()))
+        }) {
+            // If a scan range exists that contains the Orchard init height, split it in two at the
+            // init height.
+            if let Some((start, end, range_priority)) = transaction
```

Comment:
If I'm reading these correctly, then AFAICT we will _never_ see any scan range above the range that contains `orchard_init_height`, because the range we select will always be the last range. So the code below to "[r]ewrite any scanned ranges above the end of the first Orchard range to have priority `Historic`" could instead just verify that no such ranges exist (and return `CorruptedData` if they do).

Reply:
This is a result of the evolution of the migration over time. Initially, I thought that the migration should be constructed with an Orchard start height, defaulting to `max(orchard_start_height, nu5_activation)` so that existing keys with Orchard funds could have them become spendable. However, I realized that we don't actually have any existing source for such a start height other than the wallet birthday, which would trigger a full rescan, and I didn't want to do that. I came to the same conclusion that you did about the actual effect of this bit, but left it in because I figured it was safe and that we might decide to improve the decision about whether or not we're able to rescan without dropping spendability for existing notes between now and the release.